File: 157-Read-N-Characters-Given-Read4/solution.java
/* The read4 API is defined in the parent class Reader4.
      int read4(char[] buf); */

public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    public int read(char[] buf, int n) {
        if (buf == null || n <= 0) {
            return 0;
        }
        
        int i = 0;
        int tmp = 0;
        
        while (i != n) {
            tmp = read4(buf);
            if (4 == tmp) {
                i += 4;
            }
            else {
                i += tmp;
            }
        }
        
        return i;
    }
}

File: 157-Read-N-Characters-Given-Read4/solution.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        if (buf == NULL) {
            return 0;
        }
        
        int result = 0;
        int readLen = 0;

        while(result < n) {
            readLen = read4(buf + result);
            result += readLen;
            
            if (readLen != 4) {
                break;
            }
        }
        
        return min(result, n);
    }
};